### PR TITLE
CephFS: Support User and Group Quota(number of bytes) feature

### PR DESCRIPTION
**Pull request by candychencan**
**CephFS: Support User and Group Quota(number of bytes) feature**

**Summary:**
Enable individual users and groups quotas support for Cephfs.
We want to implement users and groups based quota support on individual directory hierarchies,
it means this quota distinguish users and groups, like Lustre's User、Group quota. Since currently,
Most scenarios we access Cephfs via NFS mount, but NFS exprot num has limit, So is hard to support
multiuser scenarios, but users and groups quotas can restrict the specified space for each
user and group to facilitate the fine management and application of space.

**Detailed description:**
(1) users and groups quotas can be set on any directory in the system.    
(2) users and groups quotas can restrict the **number of bytes** for the specific user or group stored 
   beneath that point in the directory hierarchy(**the number of files is not supported yet**).
(3) users and groups quotas can be set at the same time for a directory, 
   and the number of bytes reaching the quota first takes effect.

**Cephfs Modification:**
(1) add two maps:user_rbytes and group_rbytes in structure nest_info_t to store users and groups rstats.
(2) add two maps:user_max_bytes and group_max_bytes in structure quota_info_t to store users and groups quotas.
(3) during the write operation, count the number of bytes for each user and group.
(4) track size for any open file for write, and check the users and groups quota in the directory hierarchy, 
   check if the quota exceed or approaching.
(5) we can enable users or groups quotas for a directory by setfattr like:
   

> setfattr -n ceph.user_quota.max_bytes@user1 -v 1073741824 /mnt/testdir
>    setfattr -n ceph.group_quota.max_bytes@group1 -v 1073741824 /mnt/testdir

   
   we can disable users or groups quotas by setfattr like:
  

>  setfattr -n ceph.user_quota.max_bytes@user1 -v 0 /mnt/testdir
>    setfattr -n ceph.group_quota.max_bytes@group1 -v 0 /mnt/testdir

  
   we can get users or groups quotas by getfattr like:
   

> getfattr -n ceph.user_quota.max_bytes@user1  /mnt/testdir
>    getfattr -n ceph.group_quota.max_bytes@group1  /mnt/testdir

Signed-off-by: candychencan chen.can2@zte.com.cn

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by renhwztetecs** [comment]
jenkins retest this please

File: src/client/Client.cc [hunk 1]
```diff
@@ -11509,15 +11509,43 @@ int Client::_setxattr(Inode *in, const char *name, const void *value,
...
+        if (sub_name == "ceph.user_quota.max_bytes") {
+          string cp = sname.substr(sname.find("@")+1);
+          struct passwd *pw = getpwnam(cp.c_str());
+          if(!pw)
+            return -EINVAL;
+          uid_t uid = pw->pw_uid;
+          snprintf(buf, sizeof(buf),"%s@%lld", sub_name.c_str(), (long long int)uid);
+        } else if(sub_name == "ceph.group_quota.max_bytes") {
+          string cp = sname.substr(sname.find("@")+1);
+          struct group *data = getgrnam(cp.c_str());
```
Comment (ukernel) [inline src/client/Client.cc]: getting uid/gid from /etc/passwd (on the host that runs mds) is wrong.  cephfs client hosts and mds host do not have the same name->uid/gid mapping.

File: src/include/ceph_features.h [hunk 2]
```diff
@@ -248,6 +249,7 @@ DEFINE_CEPH_FEATURE_DEPRECATED(63, 1, RESERVED_BROKEN, LUMINOUS) // client-facin
 	 CEPH_FEATURE_SERVER_NAUTILUS |		\
 	 CEPH_FEATURE_CEPHX_V2 | \
 	 CEPH_FEATURE_OSD_PGLOG_HARDLIMIT | \
+	 CEPH_FEATURE_MDS_USER_GROUP_QUOTA | \
```
Comment (ukernel) [inline src/include/ceph_features.h]: add feature bit to src/mds/cephfs_features.h

File: src/mds/Server.cc [hunk 3]
```diff
@@ -5131,6 +5265,32 @@ void Server::create_quota_realm(CInode *in)
...
+  dout(10) << __func__ << " " << *in << dendl;
+
+  auto req = MClientRequest::create(CEPH_MDS_OP_SETXATTR);
+  req->set_filepath(filepath(in->ino()));
+  req->set_string2("ceph.group_quota");
+  // empty vxattr value
+  req->set_tid(mds->issue_tid());
+
+  mds->send_message_mds(req, in->authority().first);
+}
```
Comment (ukernel) [inline src/mds/Server.cc]: No reason for introducing these  functions. using Server::create_quota_realm() should be enough

File: src/client/Client.cc [hunk 4]
```diff
@@ -12612,6 +12841,22 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
...
+      return -EXDEV;
+    }
+
+    fromdir_root =
+      fromdir->quota.is_group_enable()? fromdir : get_group_quota_root(fromdir);
+    todir_root =
+      todir->quota.is_group_enable()? todir : get_group_quota_root(todir);
+    if (fromdir_root != todir_root) {
+      return -EXDEV;
+    }
```
Comment (ukernel) [inline src/client/Client.cc]: why not make quota.is_enable() cover user/group quota?

**Comment by ukernel** [comment]
My main concern is that rstat.{user_rbytes,group_rbytes} of top level directory may become very large, which may affect performance.

**Comment by renhwztetecs** [comment]
> My main concern is that rstat.{user_rbytes,group_rbytes} of top level directory may become very large, which may affect performance.

yes, when there are a lot of uids or gids, there will be some influence. We can improve the impact by replacing the existing _map_ lookup with _robin_map_, which can reduce the impact to less than 0.1%. Comparative data information will be attached later.

File: src/client/Client.cc (hunk 1 above)
Comment (candychencan) [reply src/client/Client.cc]: Thanks for your reply. In the setfattr and getfattr process,we get uid/gid from /etc/passwd on the client host and pass them to the mds host, mds host does not do mapping operation.

File: src/include/ceph_features.h (hunk 2 above)
Comment (candychencan) [reply src/include/ceph_features.h]: fixed it ,thanks.

File: src/mds/Server.cc (hunk 3 above)
Comment (candychencan) [reply src/mds/Server.cc]: fixed it ,thanks.

File: src/client/Client.cc (hunk 4 above)
Comment (candychencan) [reply src/client/Client.cc]: I have added quota.is_any_enable()  to cover directory/user/group quota, But there are scenarios where directories, users, and groups quotas need to be judged separately, So these independent judgement functions are preserved. Thanks.

**Comment by renhwztetecs** [comment]
It is interesting to use quotas based on tenants. It can extend the scope of application of CephFS. Can you give some comments?@batrick
zheng is worried about the performance problem, it does, and it only appears when a large number of tenants are set up. Later, we submit the performance data of the test for comparison.

**Comment by candychencan** [comment]
We conducted a performance comparison test. Nfs-based mounting mode, we tested the performance comparison between our user quota version and benchmark in two scenarios. The results are as follows:
image



image


1.According to the test results of copy large files, the performance of the user quota version (with no user quota set, only capacity statistics) is about 2% lower than the benchmark
2. According to the test results of the vdbench tool, the performance of the user quota version (with no no user quota set, only capacity statistics) is less than 1% lower than the benchmark.
3.Robinmap does not significantly improve performance, because when viewed with perf tools, user quota map operations are not a bottleneck in the call chain.

**Comment by candychencan** [comment]
jenkins retest this please

**Comment by candychencan** [comment]
Hi @batrick @ukernel @tchaikov Could you have time to give some comments? Thank you.

**Comment by batrick** [comment]
We'll try to take a fresh look at this in the coming month. Sorry it's been sitting in the backlog.

**Pull request closed without merging**